src/session/session_api.c

```diff
@@ -1594,14 +1598,43 @@ __wt_session_range_truncate(
      */
     if (stop != NULL) {
         WT_ERR(start->compare(start, stop, &cmp));
-        if (cmp > 0)
+        if (cmp > 0) {
+            log_trunc = true;
             goto done;
+        }
     }
 
     WT_ERR(
       __wt_schema_range_truncate(session, start, stop, orig_start_key, orig_stop_key, local_start));
 
 done:
+    /*
+     * In the cases where truncate doesn't have work to do, we still need to generate a log record
+     * for the operation. That way we can be consistent with other competing inserts or truncates on
+     * other tables in this transaction.
+     */
+    if (log_trunc) {
+        /*
+         * If we have cursors and know there is no work to do, there may not be a dhandle. Grab it
+         * from the start or stop cursor as needed.
+         */
+        dhandle = session->dhandle;
+        if (dhandle == NULL && start != NULL) {
+            cbt = (WT_CURSOR_BTREE *)start;
+            dhandle = cbt->dhandle;
+        } else if (dhandle == NULL && stop != NULL) {
+            cbt = (WT_CURSOR_BTREE *)stop;
+            dhandle = cbt->dhandle;
+        }
+        /* We have to have a dhandle from somewhere. */
+        WT_ASSERT(session, dhandle != NULL);
+        WT_WITH_DHANDLE(session, dhandle, log_op = __wt_log_op(session));
+        if (log_op) {
+            WT_WITH_DHANDLE(session, dhandle,
+              ret = __wt_txn_truncate_log(session, orig_start_key, orig_stop_key, local_start));
```

Comment:
Are there cases where an identical request would log a truncate here but *not* log a truncate if there is work for truncate to do? (Not sure if it would be a problem, since we would, I think, just be logging a no-op.)

Specifically, I'm wondering if there are paths through `__wt_schema_range_truncate()` that would not result in a logging the truncate. For example, based on a few minutes of skimming through the code, it looks like calling `__wt_schema_range_truncate()` on the history store would result in iteratively removing (and logging) individual entries from the HS.

That said, I'm not sure whether/how we can wind up truncating the HS.

Possibly I'm just worrying about something impossible or irrelevant. I'm certainly not expert in the truncate code. Maybe @agorrod will have a more useful perspective here.

Reply:
There is a test failure, which might factor into the answer here. I hadn't really thought about it, but logging the truncate here, will mean that in the simple schema case there will be two truncate log records generated per handle (I think). That's probably OK, since those log records *should* be idempotent, but there is interesting new behavior here.
OTOH: Format is currently doing a great job of letting us know when we get this wrong - so I think we can lean on testing to give us confidence.

Reply:
We are only in this code if we come through the code where we have done a `goto done` and have not called `wt_schema_range_truncate`. 

(I see the test failure and need to investigate it)